gnocchi/amqp1d.py

```diff
@@ -64,14 +64,17 @@ def _flush(self):
         for host_id, measures_by_names in six.iteritems(self._measures):
             resource = resources[host_id]
 
-            names = set(measures_by_names.keys())
+            mbn_keys = measures_by_names.keys()
+            names = (set(mbn_keys) if len(mbn_keys)
+                     else set())
```

Comment:
I'm a bit confused about this, why is this needed?

Reply:
This has something to do with the "in" operator here https://github.com/gnocchixyz/gnocchi/blob/e8df5eaa0faea8a54a3a39d859f7f9b349b52e64/gnocchi/indexer/sqlalchemy.py#L1239

This is the traceback without the above code. Actually it seems like the simple list() conversion below is enough to solve it.

```
Captured logs:
~~~~~~~~~~~~~~
    b'2021-06-29 09:50:26,837 [208372] INFO     gnocchi.amqp1d: Created resource type collectd_amqp1d'
    b'2021-06-29 09:50:26,891 [208372] ERROR    gnocchi.amqp1d: Unepected error during flush()'
    b'Traceback (most recent call last):'
    b'  File "/home/ubuntu/gnocchi/gnocchi/indexer/sqlalchemy.py", line 1401, in build_filter'
    b'    op = cls.multiple_operators[operator]'
    b"KeyError: 'in'"
    b''
    b'During handling of the above exception, another exception occurred:'
    b''
    b'Traceback (most recent call last):'
    b'  File "/home/ubuntu/gnocchi/gnocchi/amqp1d.py", line 56, in flush'
    b'    self._flush()'
    b'  File "/home/ubuntu/gnocchi/gnocchi/amqp1d.py", line 76, in _flush'
    b'    known_metrics = self.indexer.list_metrics(attribute_filter={'
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/oslo_db/api.py", line 154, in wrapper'
    b'    ectxt.value = e.inner_exc'
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/oslo_utils/excutils.py", line 227, in __exit__'
    b'    self.force_reraise()'
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/oslo_utils/excutils.py", line 200, in force_reraise'
    b'    raise self.value'
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/oslo_db/api.py", line 142, in wrapper'
    b'    return f(*args, **kwargs)'
    b'  File "/home/ubuntu/gnocchi/gnocchi/indexer/sqlalchemy.py", line 799, in list_metrics'
    b'    attribute_f = QueryTransformer.build_filter('
    b'  File "/home/ubuntu/gnocchi/gnocchi/indexer/sqlalchemy.py", line 1412, in build_filter'
    b'    return cls._handle_multiple_op(engine, table, op, nodes)'
    b'  File "/home/ubuntu/gnocchi/gnocchi/indexer/sqlalchemy.py", line 1319, in _handle_multiple_op'
    b'    return op(*['
    b'  File "/home/ubuntu/gnocchi/gnocchi/indexer/sqlalchemy.py", line 1320, in <listcomp>'
    b'    cls.build_filter(engine, table, node)'
    b'  File "/home/ubuntu/gnocchi/gnocchi/indexer/sqlalchemy.py", line 1411, in build_filter'
    b'    return cls._handle_binary_op(engine, table, op, nodes)'
    b'  File "/home/ubuntu/gnocchi/gnocchi/indexer/sqlalchemy.py", line 1391, in _handle_binary_op'
    b'    return op(attr, value)'
    b'  File "/home/ubuntu/gnocchi/gnocchi/indexer/sqlalchemy.py", line 1265, in _operator_in'
    b'    return field_name.in_(value)'
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/sqlalchemy/sql/operators.py", line 604, in in_'
    b'    return self.operate(in_op, other)'
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/sqlalchemy/orm/attributes.py", line 316, in operate'
    b'    return op(self.comparator, *other, **kwargs)'
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/sqlalchemy/sql/operators.py", line 1395, in in_op'
    b'    return a.in_(b)'
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/sqlalchemy/sql/operators.py", line 604, in in_'
    b'    return self.operate(in_op, other)'
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/sqlalchemy/orm/properties.py", line 424, in operate'
    b'    return op(self.__clause_element__(), *other, **kwargs)'
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/sqlalchemy/sql/operators.py", line 1395, in in_op'
    b'    return a.in_(b)'
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/sqlalchemy/sql/operators.py", line 604, in in_'
    b'    return self.operate(in_op, other)'
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/sqlalchemy/sql/elements.py", line 841, in operate'
    b'    return op(self.comparator, *other, **kwargs)'
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/sqlalchemy/sql/operators.py", line 1395, in in_op'
    b'    return a.in_(b)'
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/sqlalchemy/sql/operators.py", line 604, in in_'
    b'    return self.operate(in_op, other)'
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/sqlalchemy/sql/type_api.py", line 76, in operate'
    b'    return o[0](self.expr, op, *(other + o[1:]), **kwargs)'
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/sqlalchemy/sql/default_comparator.py", line 155, in _in_impl'
    b'    seq_or_selectable = coercions.expect('
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/sqlalchemy/sql/coercions.py", line 186, in expect'
    b'    resolved = impl._literal_coercion('
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/sqlalchemy/sql/coercions.py", line 564, in _literal_coercion'
    b'    self._raise_for_expected(element, **kw)'
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/sqlalchemy/sql/coercions.py", line 282, in _raise_for_expected'
    b'    util.raise_(exc.ArgumentError(msg, code=code), replace_context=err)'
    b'  File "/home/ubuntu/gnocchi/.tox/py38-mysql-file/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 207, in raise_'
    b'    raise exception'
    b'sqlalchemy.exc.ArgumentError: IN expression list, SELECT construct, or bound parameter object expected, got "{\'interface-ens2f1@if_errors-rx\', \'interface-ens2f1@if_errors-tx\', \'memory@memory-free\'}".'
    b''
```

Reply:
Should be better now right? Please let me know if you have any more feedback and if you think the approach is backportable to stable 4.4 and 4.3